### ISSUE TITLE
Entity re-parenting

### ISSUE THREAD

**Issue by caseyyee** [aframevr/aframe#2425]
It's helpful to be able to entities to different parents.   Currently, to do this, you need to:

1) flush entity to DOM (to preserve your entity state).
2) copy the entity.
3) remove the existing entity.
4) add copy to new parent entity element.

```js
var entity = document.querySelector('#myelement');
var newParent = document.querySelector('#mynewparent');
entity.flushToDom();
var copy = entity.cloneNode();
newParent.appendChild(copy);
entity.parentNode.removeChild(entity);
```

To do this in three.js, it's a simple matter of changing the `object3d.parent`.   DOM `parentNode` and `parentElement` are read only properties.  Not sure if this applies any differently to custom element/web component.

Also welcome any suggestions for any alternative way of doing this that I may have missed.

**Comment by dmarcos** [comment]
@caseyyee what are the use cases for reparenting you have in mind?

**Comment by ngokevin** [comment]
There are `Entity.add` and `Entity.addToParent`, but then they wouldn't be represented in the DOM. Maybe those methods should check if the DOM node is already attached.

Use case I can immediately think of is grabbing an object and putting it in a hand's local coordinate space.

**Comment by caseyyee** [comment]
Yes, grabbing and object would be a prime example of why you would want to do this.

The add method would work, but I figure you would want to keep the DOM tree faithful to the THREE scene structure.

**Comment by dmarcos** [comment]
We might want to wrap the methods `insertBefore` / `insertAfter` for entities

**Comment by ngokevin** [comment]
I think if we make `cloneNode` automatically `flushToDOM`, then re-parenting would be mirror normal web development, and that'd be a minimal solution.

**Comment by fernandojsg** [comment]
I guess the quick&dirty way could works but it will recreate all the components from the cloned entity.
Ideally the three.js object could be just moved using `.parent = blabla` but the DOM moved without reinitliazing the whole stuf, but I'm not sure how to do that.
I agree that a very common use will be grabbing objects and doing just `.parent = blabla` it's the way you do typically in unity for example, as it's easy and performant.

**Comment by ngokevin** [comment]
Trying to re-parent now to grab something. API Options:

- Moving a child to another parent: `.parentNode = anotherEl`
- Having a parent adopt a child: Wrap `.appendChild` to see if it's already attached, then re-parent.
- `.setParent(parentEl)`

Implementation options:

What component data do we want to keep? Do you keep local transforms? Just take original component data? Runtime data?

I might just keep it simpler and do @caseyyee's flushToDOM and copy over. Then up to application to do any processing of data.

**Comment by ngokevin** [comment]
First, a simpler solution to re-parent. Doesn't take much work so I don't think we need an API. Although we can document it.

```js
entity.flushToDom();
newParent.appendChild(entity);
```

Second, @dmarcos came up with a much better idea for temporary grabbing. Move the three.js `Object3D` as a child of the hand's `Object3D`, and leave the DOM as is. When you ungrab, you can resync the DOM if wanted. I'm gonna try that out.

**Comment by ngokevin** [comment]
Example code for grabbing / ungrabbing by re-parenting the Object3D:

Grab:

```js
 // World to local transform so position and rotation do not change when moved.
handEl.object3D.updateMatrixWorld();
worldToLocal = new THREE.Matrix4().getInverse(handEl.object3D.matrixWorld);
object3D.applyMatrix(worldToLocal);
```

Ungrab:

```js
// Get world transforms.
position = activePrimitiveObject3D.getWorldPosition();
rotation = activePrimitiveObject3D.getWorldRotation();

// Move object back to original parent.
sceneEl.object3D.add(object3D);

// Sync world transforms back to entity.
el.setAttribute('position', position);
el.setAttribute('rotation', {
  x: THREE.Math.radToDeg(rotation.x),
  y: THREE.Math.radToDeg(rotation.y),
  z: THREE.Math.radToDeg(rotation.z)
});
```

**Comment by machenmusik** [comment]
isn't flushToDOM() expensive?

**Comment by machenmusik** [comment]
you should be able to move something without owning it, personally i think reparenting should be reserved for more durable reasons.  i think of it more like tractor beam unless you are really grabbing to wield, not grab-and-drop.  should be able to do tractor beam with 3DOF, if you are moving like in editor not purely physical analog, you want to be able to push and pull much farther than 6DOF range of motion, so trackpad/thumbstick anyway

**Comment by machenmusik** [comment]
resyncing by using unmodified world position / rotation would only work if no new parent has nonzero offsets, right?

**Comment by caseyyee** [comment]
I agree with @machenmusik.

There are many general cases where you want the change reflected into the DOM and not just changed on Object3D.

Maybe a more relatable example:
```html
<a-bag></a-bag>
<a-car></a-car>
```
Now put the bag into the car and drive it around:
``` html
<a-car>
  <a-bag></a-bag>
</a-car>
```

So to answer @ngokevin earlier questions:

* `setParent(newParent)` seems to fit better in terms of consistency with other API that we already have.  I feel `appendChild()` is slightly less clear.
* Ideally we keep all components and state.   I know this is harder to achieve and for now, `flushToDom()` would probably do the trick for now.
* I wouldn't make any local/global coordinate.  Though as an option would be super nice ie:
`entity.setParent(newParent, { convertToLocal: true })`

A high effort, but maybe worthwhile  idea would be to separate the DOM tree from three.js scene graph.  Doing this, would make reparenting non-destructive and give us flexibility and performance in other areas as well.  Anyways, more pie in the sky than anything else.

**Comment by ngokevin** [comment]
Yeah, a nice to have. But working around or doing it manually is fairly easy. If I wanted to add DOM syncing or account for controller offsets, I could.

**Comment by scott-plutovr** [comment]
We have an a-entity with a number of children that we are trying to re-parent to the left or right Vive controller based on the menu pressed. When doing so we receive an TypeError from index.js:270 that states, "position is undefined"

We have tried the flushToDOM() call on the entity prior to the appendChild() that you recommend above. We have also tried the initial list of stuff above where we cloneNode() prior to appendChild() followed by removeChild(). Nothing has worked.

Here's the call stack if it helps,
 `TypeError: position is undefined[Learn More]
index.js:270
setPositions/<
index.js:270
forEach self-hosted:281:13 setPositions
index.js:267
remove
index.js:101
value
a-entity.js:428:6
[161]</module.exports/</<
bind.js:12:13
forEach self-hosted:281:13 value
a-entity.js:106:6
m
document-register-element.js:2:1464
l
document-register-element.js:2:80
o/<
document-register-element.js:2:210
r
document-register-element.js:2:4796
[16]</</r[a]/O</<
document-register-element.js:2:4917`

**Comment by roskelld** [comment]
I've been working on a setup for re-parenting objects, including the camera. I want to do this to utilise the transform updates of the parent, but maintaining the worldspace of the child object as it is parented and unparented.

Use cases: an elevator or tray of objects. 

Keeping the worldspace transform and rotation of the child would allow a camera to step on an elevator, move freely whilst on it, then step off on a different floor seamlessly.

I found the following code works great for non-kinematic type objects. 

```
          child.object3D.parent.updateMatrixWorld();
          // remove child from scene and add it to parent
          THREE.SceneUtils.attach( child.object3D, child.sceneEl.object3D, parent.object3D );
```

When I test the code out with the camera, however, there's a transform jump as I transition on or off the parent object. It looks like the camera worldtransform is stuck at the initial spawn point. 

CAMERA
[BEFORE]
Child Position (world) Vector3 { x: 1.11, y: 1.6, z: -17.91 }
Old Parent Position Vector3 { x: 0, y: 0, z: 0 }
[AFTER]
New Parent Position Vector3 { x: 0, y: -0.14, z: -20 }
Child Position (world) Vector3 { x: 1.11, y: 1.45, z: -37.94 }
Child Position (local) Vector3 { x: 1.11, y: 1.6, z: -17.94 }

STATIC OBJECT
[BEFORE]
Child Position (world) Vector3 { x: 0, y: 5, z: -12 }
Old Parent Position Vector3 { x: 0, y: 0, z: 0 }
[AFTER]
New Parent Position Vector3 { x: 15, y: 0, z: -3 }
Child Position (world) Vector3 { x: 0, y: 5, z: -12 }
Child Pposition (local) Vector3 { x: -15, y: 5, z: -9 }

Any thoughts on how to resolve this?

**Comment by ngokevin** [comment]
For reference, the current mode of simulating "grabbing" or re-parenting: https://www.youtube.com/watch?v=Gd-Qc9TuHZ0

Umm off top of my head, I think it's like: have a dummy object under the parent. Invert the entity's world matrix/transforms. Apply dummy object's matrixWorld transforms to entity. Or something similar. If I get around to this again, I'll post code.

**Comment by PlumCantaloupe** [comment]
What about some sort of temporary parent-constraint for when we don't want to mess with the scene hierarchy, inspired by [Maya](https://knowledge.autodesk.com/support/maya/learn-explore/caas/CloudHelp/cloudhelp/2016/ENU/Maya/files/GUID-2FC0EE9E-A8D5-4B2A-A751-8B926389497A-htm.html)?

Here is an example component [on Glitch]( https://glitch.com/edit/#!/aframe-parent-constraint ). I haven't battle tested it yet but perhaps a start?

```javascript 
AFRAME.registerComponent('mdmu-parent-constraint', {
            //dependencies: ['mdmu-gltf', 'mdmu-controls'],
            schema: {
                parent:     {type: "selector", default:null},
                position:   {type: "boolean",  default:true}, //default true to update all transforms
                rotation:   {type: "boolean",  default:true},
                scale:      {type: "boolean",  default:true}
            },
            multiple: false, //do not allow multiple instances of this component on this entity
            init: function() {
                this.psuedoParent = null;
                this.originalLocalTransform = this.el.object3D.matrix.clone(); //save original local transform to use later
            },
            update: function(oldData)  {
                const data = this.data;

                if (Object.keys(data).length === 0) { return; } // No need to update. as nothing here yet

                //model change
                if ( (oldData.parent !== data.parent) && data.parent !== null ) {
                    this.psuedoParent = document.querySelector('#' + data.parent.getAttribute('id'));
                    this.el.object3D.matrixAutoUpdate = false; //we want to manually update here and not have conflicts elsewhere
                }

                //remove this component if you want to break constraint .... setting to null will do nothing
            },
            tick: function(time, timeDelta) {
                if (this.psuedoParent !== null) {
                    const Context_AF    = this;
                    const data          = Context_AF.data;

                    let parentObject3D          = Context_AF.psuedoParent.object3D;
                    let thisObject3D            = Context_AF.el.object3D;
                    let worldMat_Constraint     = new THREE.Matrix4();
                    let worldMat_NoConstraint   = new THREE.Matrix4();
                    //console.log( worldMat_Constraint );
                    let position_P              = new THREE.Vector3();
                    let position_C              = new THREE.Vector3();
                    let rotation_P              = new THREE.Quaternion();
                    let rotation_C              = new THREE.Quaternion();
                    let scale_P                 = new THREE.Vector3();
                    let scale_C                 = new THREE.Vector3();

                    //get world matrix of pseudo-parent we want to constrain to
                    worldMat_Constraint.copy( parentObject3D.matrixWorld ); 

                    //get world matrix of this object if we didn't apply a constraint
                    worldMat_NoConstraint.copy( thisObject3D.parent.matrixWorld );         
                    worldMat_NoConstraint.premultiply( Context_AF.originalLocalTransform );  
                    
                    console.log( worldMat_NoConstraint.getPosition() );

                    //break down into individual transforms ... thanks for the handy function THREEjs!
                    worldMat_Constraint.decompose(position_P, rotation_P, scale_P);
                    worldMat_NoConstraint.decompose(position_C, rotation_C, scale_C); //apply saved local transform

                    //if we want not constrain one of these transforms we will "reset" it back to what it would be with no constraint applied
                    if (!data.position) {
                        position_P.copy(position_C);
                    }

                    if (!data.rotation) {
                        rotation_P.copy(rotation_C);
                    }

                    if (!data.scale) {
                        scale_P.copy(scale_C);
                    }

                    //recompose world matrix with adjusted transforms
                    worldMat_Constraint.compose(position_P, rotation_P, scale_P);   

                    //convert to local matrix of this object3D
                    const thisLocalInverse = new THREE.Matrix4();
                    thisLocalInverse.getInverse( thisObject3D.parent.matrixWorld );
                    worldMat_Constraint.premultiply( thisLocalInverse );

                    //set new matrix and manually update
                    thisObject3D.matrix.copy(worldMat_Constraint);                         
                    thisObject3D.updateMatrixWorld(true);  
                }
            },
            remove: function() {
                this.el.object3D.matrixAutoUpdate = true;   //give back control of update to framework
            }
        }); 
```

**Comment by damned** [comment]
bump with reasoning:

this feels to me a disproportionately troublesome and tricky problem to code around, especially for new aframe or 3d developers (like me), for a need that seems likely very common.

workarounds to vary the three object graph away from the a-frame entity graph, or use tracking mechanisms or manually deal with scale and rotations individually are great to get us through, but are:
 - not robust
 - break obvious core patterns (entity hierarchy, entity<->object3D relationship)
 - create extra complications and special cases that are likely going to get people into more problems down the line

use case seems pretty clear:
 - move entity within hierarchy
 - by default retain its world position

many of the chunks of code we have poking under the covers just no longer work - in part or in whole - so, having versioned code maintained as part of aframe is imho pretty much a necessity

---

that said, to add to the soon-to-be-out-of-date snippets, here's my contribution :)

it gives reparenting **for single entities** while maintaining worldspace presence (including scale) which *seems to* work for my use cases
aframe 1.0.4 
oculus quest v18

```
let matrixCopy = (matrix) => new THREE.Matrix4().copy(matrix)
const identityMatrix = new THREE.Matrix4().identity()

function reparent(entity, newParent, done) {
  let intendedWorldMatrix = matrixCopy(entity.object3D.matrixWorld) // this encodes starting pos, rot and scale of entity being moved
  entity.parentElement.removeChild(entity)

  let reparented = entity.cloneNode()
  newParent.appendChild(reparented)

  reparented.addEventListener('loaded', () => {
    // based on parent-world.local matrix multiplication order for child's world matrix:
    //   https://github.com/mrdoob/three.js/blob/dev/src/core/Object3D.js#L560
    //
    // determine required child local matrix thus:
    //   https://math.stackexchange.com/questions/949341/how-to-find-matrix-b-given-matrix-ab-and-a          
    let recalculatedLocalMatrix = new THREE.Matrix4().getInverse(newParent.object3D.matrixWorld).multiply(intendedWorldMatrix)

    let matrixAutoUpdate = reparented.object3D.matrixAutoUpdate
    reparented.object3D.matrixAutoUpdate = false // so that local matrix doesn't get trashed on tick while updating 
    reparented.object3D.matrix.copy(recalculatedLocalMatrix)
    reparented.object3D.applyMatrix(identityMatrix) // to set pos, rot, scale from matrix
    reparented.object3D.matrixAutoUpdate = matrixAutoUpdate
    newParent.object3D.updateMatrixWorld() // ensure all world matrices around moved entity are consistent

    done()
  })
  return reparented
}
```
FYI aframe 1.0.4 uses three.js r111, so applyMatrix is used, instead of its replacement (applyMatrix4) at three.js head

NB reparenting entities with children of their own would require a little more work - clone recursively (`cloneNode(true)`) and deal with resetting their matrices as intended (although simpler as only copy from previous local matrices required - i.e. no back-calculation - followed by applyMatrix(identity) on each before updateMatrixWorld() called)

**Comment by jgbarah** [comment]
I've explored a different solution, based on using `object3D.attach` to get the coordinates with respect to the intended parent, and then cloning the entity (and later removing the "old" one). See a component implemented following that approach in  [this answer to "AFrame: reparenting an element keeping its world position, rotation"](https://stackoverflow.com/a/65554657/2075265).
